File: 5-valid-parentheses.py
def isValidParantheses(string):
    parens = {"(": ")", "{": "}", "[": "]"}
    stack = []

    # Check the character of string
    # If the character is one of the opening parentheses, put it in the stack
    # If the character is one of the closing parentheses, check if that is closing bracket of last parentheses in the stack
    # If the stack is empty or the last opening parentheses and the last closing parentheses don't match, return False
    # Or return True

    # 0(n) time | 0(n) space

    for character in string:
        if character in parens:  # opening
            stack.append(character)
        else:  # closing
            if not stack or character != parens[stack.pop()]:
                # parens[stack.pop() -> stack:({[ character :}
                # or stack is empty
                return False
    return not stack


print(isValidParantheses("{([])}"))
print(isValidParantheses("{(})"))

File: 5-valid-parentheses.js
function isValidParentheses(string) {
	const parens = { '(': ')', '{': '}', '[': ']' };
	const stack = [];

	for (const [index, character] of Array.from(string).entries()) {
		console.log(index, character);
		if (Object.hasOwn(parens, character)) {
			stack.push(character);
		} else {
			if (stack.length < 1 || character !== parens[stack.pop()]) {
				return false;
			}
		}
	}
	return stack.length < 1;
}

console.log(isValidParentheses('{([])}'));
console.log(isValidParentheses('{(})'));
